proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1097 @@
+   because it rejects expressions like ::
+
+     ((\x -> (x :: a)) :: forall a. a -> a)
+
+   which are accepted today. No migration period is proposed, because it is
+   very hard to imagine how ``-XTypeAbstractions`` and ``-XScopedForAlls`` should
+   co-exist peacefully here. Instead, we can issue a specific error message telling
+   users how to migrate their code in this case.
+
+   My hope is that constructs such as this one are rare and would not impact many
+   users.
+
+   If necessary, we could imagine taking the expression ``expr :: forall ... . ty``
+   and looking proactively to see whether ``expr`` ever uses a type variable
+   pattern from this proposal. If not, ``-XScopedForAlls`` could trigger (and we
+   issue a warning with ``-Wcompat``). But, if a type argument appears anywhere
+   in ``expr``, then ``-XScopedForAlls`` is disabled. This would be backward-compatible,
+   but unfortunately non-local and annoying. I prefer just to skip this
+   migration step.
+
+``let``-binding types
+~~~~~~~~~~~~~~~~~~~~~
+
+.. _type-let:
+
+This segment of the proposal goes beyond previous proposals in describing a mechanism
+to use ``let`` to bind type synonyms.
+
+Motivation
+^^^^^^^^^^
+
+1. Users have, from time to time, requested the ability to make local type synonyms.
+   GHC even has a little support for synonyms via equality constraints (e.g., writing
+   ``f :: (a ~ Some Big Type With Lots Of Parts) => Maybe a -> a -> Maybe a``). Instead
+   of encoding this idea via equality constraints, though, it would be nice to support
+   it directly.
+
+#. Now that type variables can stand for types, we can write code like ::
+
+     f :: Maybe Bool -> Bool
+     f (x :: Maybe b) = (True :: b)
+
+   Note that the pattern signature binds ``b`` to ``Bool``. This is, essentially, a ``let``\ -bound
+   type variable: in the scope of ``b``, ``b`` is synonymous with ``Bool``. Yet the only way
+   to make such a ``b`` is via a pattern (or result, `#228`_) signature. Why force users
+   to use matching instead of binding the variable directly.
+
+#. Doing this helps uphold the EBP_.
+
+Specification
+^^^^^^^^^^^^^
+
+1. Create a new extension ``-XExtendedLet``.
+
+#. With ``-XExtendedLet``, add two new productions for ``decl`` (from the `Haskell 2010 Report`_), ::
+
+     decl → 'type' simpletype '=' type
+          → 'type' tyvar '=' type
+
+   and remove the production ``topdecl → 'type' simpletype '=' type`` from ``topdecl``.
+
+   Note that the second form allows a local binding for a lower-case ``tyvar``; these
+   synonyms may not be parameterized.
+
+#. The form ``decl → 'type' tyvar '=' type`` is not allowed at top-level.
+
+#. These new declaration forms introduce local type synonyms in terms, which scope over the same
+   region of code that other declarations in the same ``let`` / ``where`` clause scope over.
+
+   Like other type synonyms, local type synonyms may not be recursive.
```

Comment:
I think that you want to be more specific about what the semantics of both forms of local type synonyms is. In particular when they appear in a let block alongside other type synonyms, and regular let bindings.

Why do we need the lower-case form? Apart from “because we can” I don't see much justification.

Reply:
Actually, the `let` for type variables doesn't uphold the EBP on its own. So I moved that text closer to here, where it now also answers the question about lower-case variables.